src/neo/Network/P2P/RemoteNode.ProtocolHandler.cs

```diff
@@ -270,15 +272,15 @@ private void OnGetDataMessageReceived(InvPayload payload)
         /// Tell the specified number of blocks' headers starting with the requested IndexStart to RemoteNode actor.
         /// A limit set by HeadersPayload.MaxHeadersCount is also applied to the number of requested Headers, namely payload.Count.
         /// </summary>
-        /// <param name="payload">A GetBlocksPayload including start block index and number of blocks' headers requested.</param>
+        /// <param name="payload">A GetBlockByIndexPayload including start block index and number of blocks' headers requested.</param>
         private void OnGetHeadersMessageReceived(GetBlockByIndexPayload payload)
         {
             uint index = payload.IndexStart;
             if (index > Blockchain.Singleton.Height) return;
             List<Header> headers = new List<Header>();
             using (SnapshotCache snapshot = Blockchain.Singleton.GetSnapshot())
             {
-                uint count = payload.Count == -1 ? HeadersPayload.MaxHeadersCount : (uint)payload.Count;
+                uint count = payload.Count == -1 || payload.Count > HeadersPayload.MaxHeadersCount ? HeadersPayload.MaxHeadersCount : (uint)payload.Count;
```

Comment:
```suggestion
                uint count = payload.Count == -1 ? HeadersPayload.MaxHeadersCount : (uint)payload.Count;
```

If `payload.Count > HeadersPayload.MaxHeadersCount`, if throws in `GetBlockByIndexPayload.Deserialize()`.